devtools/reflection-agent/src/main/java/io/quarkus/agent/ReflectionAgent.java

```diff
@@ -0,0 +1,121 @@
+package io.quarkus.agent;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.lang.instrument.ClassFileTransformer;
+import java.lang.instrument.IllegalClassFormatException;
+import java.security.ProtectionDomain;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+
+public class ReflectionAgent {
+
+    private static volatile Set<String> KNOWN_CLASSES;
+
+    private static final Set<String> warned = Collections.newSetFromMap(new ConcurrentHashMap<>());
+
+    public static void premain(java.lang.String s, java.lang.instrument.Instrumentation i) {
+        i.addTransformer(new ClassFileTransformer() {
+            @Override
+            public byte[] transform(ClassLoader loader, String className, Class<?> classBeingRedefined,
+                    ProtectionDomain protectionDomain, byte[] classfileBuffer) throws IllegalClassFormatException {
+                if (className.startsWith("io/quarkus/agent")
+                        || className.startsWith("java/")
+                        || className.startsWith("javax/")
+                        || className.startsWith("jdk/")
+                        || className.startsWith("sun/")
+                        || className.startsWith("org/jboss/log")) {
+                    return null;
+                }
+                try {
+                    AtomicBoolean modified = new AtomicBoolean(false);
+                    ClassReader reader = new ClassReader(className);
+                    ClassWriter writer = new ClassWriter(reader, Opcodes.ASM6);
+                    reader.accept(new ClassVisitor(Opcodes.ASM6, writer) {
+                        @Override
+                        public MethodVisitor visitMethod(int access, String name, String descriptor, String signature,
+                                String[] exceptions) {
+                            MethodVisitor existing = super.visitMethod(access, name, descriptor, signature, exceptions);
+                            return new MethodVisitor(Opcodes.ASM6, existing) {
+                                @Override
+                                public void visitMethodInsn(int opcode, String owner, String name, String descriptor,
+                                        boolean isInterface) {
+                                    if (owner.equals("java/lang/Class") && name.equals("forName")) {
```

Comment:
Could we also try to get `getDeclaredFields()` and `getDeclaredMethods()` as this is an issue for reflection? If we call these methods, all the fields, respectively methods needs to be registered for reflection, otherwise you have partial results.

Reply:
We could, but for now I only needed this one. Its a bit more complex to do methods/fields so I am not sure when I will get to it

Reply:
OK, I'll create an issue so that we don't forget about it.